server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -370,7 +364,24 @@ protected int getRefCount(ElasticsearchDirectoryReader reference) {
         }
 
         @Override
-        protected void decRef(ElasticsearchDirectoryReader reference) throws IOException { reference.decRef(); }
+        protected void decRef(ElasticsearchDirectoryReader reference) throws IOException {
+            reference.decRef();
+        }
+    }
+
+    @Override
+    final boolean assertSearcherIsWarmedUp(String source, SearcherScope scope) {
+        if (scope == SearcherScope.EXTERNAL) {
+            switch (source) {
+                // we can access segment_stats while a shard is still in the recovering state.
+                case "segments":
+                case "segments_stats":
+                    break;
+                default:
+                    assert externalReaderManager.isWarmedUp : "searcher was not warmed up yet for source[" + source + "]";
```

Comment:
This assert seems to access the boolean unsafely. Not guarded by refreshLock here AFAICS.

Reply:
Yeah, the comment is outdated with the assertion. I think we have happens-before relationships in IndexShard (in postRecovery and resetEngineToGlobalCheckpoint). Are you okay to leave it as is?

Reply:
Looks good, thanks for clarifying.